File: api/app/schemas/insights.py
from app.schemas.base import CamelModel


class GroupBreakdownItem(CamelModel):
    group_id: str
    group_name: str
    group_color: str | None
    minutes: int
    previous_minutes: int
    minutes_delta: int
    percent_change: float
    percent_of_total: float


class TopCategoryItem(CamelModel):
    category_id: str
    category_name: str
    group_name: str
    group_color: str | None
    minutes: int
    previous_minutes: int
    minutes_delta: int
    percent_change: float
    percent_of_total: float


class ProductiveDayDetail(CamelModel):
    date: str
    day_name: str
    minutes: int


class LongestActivityDetail(CamelModel):
    category_name: str
    minutes: int
    date: str
    start_time: str
    end_time: str


class WeeklyStats(CamelModel):
    activities_count: int
    previous_activities_count: int
    activities_count_delta: int
    categories_used: int
    previous_categories_used: int
    categories_used_delta: int
    average_activity_duration: float
    previous_average_activity_duration: float
    average_activity_duration_delta: float
    average_daily_minutes: float
    previous_average_daily_minutes: float
    average_daily_minutes_delta: float
    most_productive_day: ProductiveDayDetail | None
    least_productive_day: ProductiveDayDetail | None
    longest_activity: LongestActivityDetail | None


class DailyBreakdownItem(CamelModel):
    date: str
    day_name: str
    minutes: int
    activities_count: int


class GoalProgressItem(CamelModel):
    category_id: str
    category_name: str
    current_week_minutes: int
    target_weekly_minutes: int
    min_weekly_minutes: int
    max_weekly_minutes: int
    progress_percent: float
    status: str


class WeeklyComparisonResponse(CamelModel):
    week_start_date: str
    week_end_date: str
    previous_week_start_date: str
    previous_week_end_date: str
    total_minutes: int
    previous_total_minutes: int
    total_minutes_delta: int
    total_minutes_percent_change: float
    group_breakdown: list[GroupBreakdownItem]
    top_categories: list[TopCategoryItem]
    stats: WeeklyStats
    daily_breakdown: list[DailyBreakdownItem]
    goals_progress: list[GoalProgressItem] | None


class DailyStats(CamelModel):
    activities_count: int
    previous_activities_count: int
    activities_count_delta: int
    categories_used: int
    previous_categories_used: int
    categories_used_delta: int
    average_activity_duration: float
    previous_average_activity_duration: float
    average_activity_duration_delta: float
    longest_activity: LongestActivityDetail | None


class ProductivityBreakdown(CamelModel):
    mandatory_minutes: int
    previous_mandatory_minutes: int
    mandatory_minutes_delta: int
    mandatory_percent_of_total: float
    optional_minutes: int
    previous_optional_minutes: int
    optional_minutes_delta: int
    optional_percent_of_total: float


class DailyComparisonResponse(CamelModel):
    date: str
    previous_date: str
    total_minutes: int
    previous_total_minutes: int
    total_minutes_delta: int
    total_minutes_percent_change: float
    group_breakdown: list[GroupBreakdownItem]
    top_categories: list[TopCategoryItem]
    stats: DailyStats
    productivity: ProductivityBreakdown | None

File: front/src/lib/api/schemas/insights.ts
import { z } from 'zod'

export const GroupBreakdownItemSchema = z.object({
  groupId: z.string().uuid(),
  groupName: z.string(),
  groupColor: z.string().nullable(),
  minutes: z.number(),
  previousMinutes: z.number(),
  minutesDelta: z.number(),
  percentChange: z.number(),
  percentOfTotal: z.number(),
})

export const TopCategoryItemSchema = z.object({
  categoryId: z.string().uuid(),
  categoryName: z.string(),
  groupName: z.string(),
  groupColor: z.string().nullable(),
  minutes: z.number(),
  previousMinutes: z.number(),
  minutesDelta: z.number(),
  percentChange: z.number(),
  percentOfTotal: z.number(),
})

export const ActivityHighlightSchema = z.object({
  categoryName: z.string(),
  minutes: z.number(),
  startTime: z.string(),
  endTime: z.string(),
})

export const DailyStatsSchema = z.object({
  activitiesCount: z.number(),
  previousActivitiesCount: z.number(),
  activitiesCountDelta: z.number(),
  categoriesUsed: z.number(),
  previousCategoriesUsed: z.number(),
  categoriesUsedDelta: z.number(),
  averageActivityDuration: z.number(),
  previousAverageActivityDuration: z.number(),
  averageActivityDurationDelta: z.number(),
  longestActivity: ActivityHighlightSchema.nullish(),
  shortestActivity: ActivityHighlightSchema.nullish(),
})

export const ProductivitySchema = z.object({
  mandatoryMinutes: z.number(),
  previousMandatoryMinutes: z.number(),
  mandatoryMinutesDelta: z.number(),
  mandatoryPercentOfTotal: z.number(),
})

export const DailyInsightSchema = z.object({
  date: z.string().date(),
  previousDate: z.string().date(),
  totalMinutes: z.number(),
  previousTotalMinutes: z.number(),
  totalMinutesDelta: z.number(),
  totalMinutesPercentChange: z.number(),
  groupBreakdown: z.array(GroupBreakdownItemSchema),
  topCategories: z.array(TopCategoryItemSchema).max(5),
  stats: DailyStatsSchema,
  productivity: ProductivitySchema.nullish(),
})

// Weekly Insights Schemas
export const DailyBreakdownItemSchema = z.object({
  date: z.string().date(),
  dayName: z.string(),
  minutes: z.number(),
  activitiesCount: z.number(),
})

export const ProductiveDaySchema = z.object({
  date: z.string().date(),
  dayName: z.string(),
  minutes: z.number(),
})

export const WeeklyActivityHighlightSchema = z.object({
  categoryName: z.string(),
  minutes: z.number(),
  date: z.string().date(),
  startTime: z.string(),
  endTime: z.string(),
})

export const WeeklyStatsSchema = z.object({
  activitiesCount: z.number(),
  previousActivitiesCount: z.number(),
  activitiesCountDelta: z.number(),
  categoriesUsed: z.number(),
  previousCategoriesUsed: z.number(),
  categoriesUsedDelta: z.number(),
  averageActivityDuration: z.number(),
  previousAverageActivityDuration: z.number(),
  averageActivityDurationDelta: z.number(),
  averageDailyMinutes: z.number(),
  previousAverageDailyMinutes: z.number(),
  averageDailyMinutesDelta: z.number(),
  mostProductiveDay: ProductiveDaySchema.nullish(),
  leastProductiveDay: ProductiveDaySchema.nullish(),
  longestActivity: WeeklyActivityHighlightSchema.nullish(),
})

export const GoalProgressSchema = z.object({
  categoryId: z.string().uuid(),
  categoryName: z.string(),
  currentWeekMinutes: z.number(),
  targetWeeklyMinutes: z.number(),
  minWeeklyMinutes: z.number(),
  maxWeeklyMinutes: z.number(),
  progressPercent: z.number(),
  status: z.enum(['under', 'on_track', 'target_met', 'over']),
})

export const WeeklyInsightSchema = z.object({
  weekStartDate: z.string().date(),
  weekEndDate: z.string().date(),
  previousWeekStartDate: z.string().date(),
  previousWeekEndDate: z.string().date(),
  totalMinutes: z.number(),
  previousTotalMinutes: z.number(),
  totalMinutesDelta: z.number(),
  totalMinutesPercentChange: z.number(),
  groupBreakdown: z.array(GroupBreakdownItemSchema),
  topCategories: z.array(TopCategoryItemSchema).max(5),
  stats: WeeklyStatsSchema,
  dailyBreakdown: z.array(DailyBreakdownItemSchema),
  goalsProgress: z.array(GoalProgressSchema).nullish(),
})

export type GroupBreakdownItem = z.infer<typeof GroupBreakdownItemSchema>
export type TopCategoryItem = z.infer<typeof TopCategoryItemSchema>
export type ActivityHighlight = z.infer<typeof ActivityHighlightSchema>
export type DailyStats = z.infer<typeof DailyStatsSchema>
export type Productivity = z.infer<typeof ProductivitySchema>
export type DailyInsight = z.infer<typeof DailyInsightSchema>
export type DailyBreakdownItem = z.infer<typeof DailyBreakdownItemSchema>
export type ProductiveDay = z.infer<typeof ProductiveDaySchema>
export type WeeklyActivityHighlight = z.infer<typeof WeeklyActivityHighlightSchema>
export type WeeklyStats = z.infer<typeof WeeklyStatsSchema>
export type GoalProgress = z.infer<typeof GoalProgressSchema>
export type WeeklyInsight = z.infer<typeof WeeklyInsightSchema>
